source/common/http/async_client_impl.cc

```diff
@@ -184,36 +198,33 @@ AsyncRequestImpl::AsyncRequestImpl(MessagePtr&& request, AsyncClientImpl& parent
 
 void AsyncRequestImpl::initialize() {
   sendHeaders(request_->headers(), !request_->body());
-  if (!remoteClosed() && request_->body()) {
-    sendData(*request_->body(), true);
+  if (request_->body()) {
+    // sendHeaders can result in synchronous stream closure in certain cases (e.g. connection pool
```

Comment:
I thought the goal here was to have `onComplete` only called in non-reset cases. Is that not true anymore? If so can we update the interface comments?

Reply:
Nope, this is still correct - only one of `onComplete` or `onReset` should ever be called for a given stream/request.

Reply:
This is a workaround to the fact that `send*` can result in synchronous remote closure _without_ it being a reset. `AsyncRequestImpl` previously fired `onComplete` as soon as it received a complete response, regardless of whether it had finished sending the request. This maintains that behavior.

I suggested above (somewhere) an alternative where the connection pool could surface a connection failure as a reset and we map that to a 503 outside the local stream. But that could also have further reaching implications.

Actually, previously, it was possible for `cleanup` to get fired multiple times for a stream (and some tests exhibited that behavior). The new checks in `closeLocal` and `closeRemote` now prevent that from happening (as well as guarantee that only one of `onComplete` or `onReset` is ever called).